src/coprocessor/dag/expr/builtin_cast.rs

```diff
@@ -359,4 +456,87 @@ impl FnCall {
     ) -> Result<Option<Vec<Duration>>> {
         unimplemented!()
     }
+
+    pub fn cast_json_as_duration(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Duration>>> {
+        unimplemented!()
+    }
+
+    pub fn cast_int_as_json(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Json>>> {
+        unimplemented!()
+    }
+
+    pub fn cast_real_as_json(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Json>>> {
+        unimplemented!()
+    }
+
+    pub fn cast_decimal_as_json(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Json>>> {
+        unimplemented!()
+    }
+
+    pub fn cast_str_as_json(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Json>>> {
+        unimplemented!()
+    }
+
+    pub fn cast_time_as_json(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Json>>> {
+        unimplemented!()
+    }
+
+    pub fn cast_duration_as_json(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Json>>> {
+        unimplemented!()
+    }
+
+    pub fn cast_json_as_json(
+        &self,
+        ctx: &StatementContext,
+        row: &[Datum],
+    ) -> Result<Option<Vec<Json>>> {
+        unimplemented!()
+    }
+
+    /// `produce_float_with_specified_tp`(`ProduceFloatWithSpecifiedTp` in tidb) produces
+    /// a new float64 according to `flen` and `decimal` in `self.tp`.
+    /// TODO port tests from tidb(tidb haven't implemented now)
+    fn produce_float_with_specified_tp(&self, ctx: &StatementContext, f: f64) -> Result<f64> {
+        let flen = self.tp.get_flen();
+        let decimal = self.tp.get_decimal();
+        if flen == convert::UNSPECIFIED_LENGTH || decimal == convert::UNSPECIFIED_LENGTH {
+            return Ok(f);
+        }
+        match convert::truncate_f64(f, flen as u8, decimal as u8) {
+            Res::Ok(d) | Res::Truncated(d) => Ok(d),
+            Res::Overflow(d) => {
+                //TODO process warning with ctx
+                try!(convert::handle_truncate(ctx, true));
```

Comment:
Since it would never return truncated in `truncate_f64` , the related code in `tidb` is https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L780 and https://github.com/pingcap/tidb/blob/master/util/types/helper.go#L71